### PR TITLE
ddl:add constraints check when creating table partition

### PR DESCRIPTION
**Pull request by ciscoxll**
## What have you changed? (mandatory)

This PR mainly solves when creating table partition, checks that the range partitioning key is included in the table constraint.
1. Returns directly if there is no constraint in the partition table.
2. Save the column names in table constraints to `[]map[string]struct{}`.
3. Checks that the partitioning key is included in the constraint.

## What is the type of the changes? (mandatory)

New feature
- New feature (non-breaking change which adds functionality)

## How has this PR been tested? (mandatory)
Unit tests.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Does this PR need to be added to the release notes? (mandatory)
No

### CODE REVIEW

**Comment by ciscoxll** [comment]
@tiancaiamao @zimulala @winkyao PTAL

File: ddl/partition.go [hunk 1]
```diff
@@ -280,3 +281,95 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+			return ErrUniqueKeyNeedAllFieldsInPf.GenByArgs("PRIMARY KEY")
+		}
+	default:
+		// Other expression, with multiple column names.
+		if checkConstraintIncludePartKey(partkeys, priKeys, true) || checkConstraintIncludePartKey(partkeys, multipleKeys, true) {
+			return ErrUniqueKeyNeedAllFieldsInPf.GenByArgs("PRIMARY KEY")
+		}
+
+	}
+	// Range partitioning key can only contain one unique key.
```
Comment (tiancaiamao) [inline ddl/partition.go]: MySQL document link ?

File: ddl/db_test.go [hunk 2]
```diff
@@ -2842,6 +2842,182 @@ func (s *testDBSuite) TestTruncatePartitionAndDropTable(c *C) {
...
+	s.testErrorCode(c, sql3, tmysql.ErrUniqueKeyNeedAllFieldsInPf)
+
+	s.tk.MustExec("drop table if exists Part1;")
+	sql4 := `create table Part1 (
+		col1 int not null,
+		col2 date not null,
+		col3 int not null,
+		col4 int not null,
+		unique key (col1, col2, col3),
+  		unique key (col3)
```
Comment (tiancaiamao) [inline ddl/db_test.go]: ident

File: ddl/partition.go [hunk 3]
```diff
@@ -280,3 +281,97 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+	}
+	cols := expression.ExtractColumns(e)
+	for _, col := range cols {
+		partkeys = append(partkeys, col.ColName.L)
+	}
+
+	switch expr.(type) {
+	case *ast.ColumnNameExpr, *ast.FuncCallExpr:
+		// Only one column name expression.
+		if checkConstraintIncludePartKey(partkeys, priKeys, false) || checkConstraintIncludePartKey(partkeys, multipleKeys, false) {
```
Comment (tiancaiamao) [inline ddl/partition.go]: partition by range (year(col1) + year(col2))

Is it `*ast.FuncCallExpr` and is there "only one column name expression" ?

File: ddl/partition.go [hunk 4]
```diff
@@ -280,3 +281,97 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+		return errors.Trace(err)
+	}
+	cols := expression.ExtractColumns(e)
+	for _, col := range cols {
+		partkeys = append(partkeys, col.ColName.L)
+	}
+
+	switch expr.(type) {
+	case *ast.ColumnNameExpr, *ast.FuncCallExpr:
+		// Only one column name expression.
```
Comment (tiancaiamao) [inline ddl/partition.go]: This make me confused.

```
partition by range (year(col1) + year(col2))
```

Is it `*ast.FuncCallExpr` and is there "only one column name expression" ?

File: ddl/partition.go [hunk 5]
```diff
@@ -280,3 +281,97 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+	// see https://dev.mysql.com/doc/refman/5.7/en/partitioning-limitations-partitioning-keys-unique-keys.html.
+	if len(uniKeys) > 0 {
+		if len(uniKeys) > 1 || !strings.Contains(partkeys[0], uniKeys[0]) {
+			return ErrUniqueKeyNeedAllFieldsInPf.GenByArgs("UNIQUE INDEX")
+		}
+	}
+	return nil
+}
+
+// saveConstraintsColumnNames save the column names in the table constraint in slice.
```
Comment (tiancaiamao) [inline ddl/partition.go]: Add comment about the return value

File: ddl/partition.go [hunk 6]
```diff
@@ -280,3 +281,97 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+		}
+	default:
+		// Other expression, with multiple column names.
+		if checkConstraintIncludePartKey(partkeys, priKeys, true) || checkConstraintIncludePartKey(partkeys, multipleKeys, true) {
+			return ErrUniqueKeyNeedAllFieldsInPf.GenByArgs("PRIMARY KEY")
+		}
+
+	}
+
+	// Range partitioning key can only contain one unique key.
```
Comment (tiancaiamao) [inline ddl/partition.go]: I don't see this restriction. 
Is it a legal one:

```
create table Part1 (
		col1 int not null,
		col2 date not null,
		col3 int not null,
		col4 int not null,
		unique key (col1, col2),
		unique key (col1, col3)
) partition by range( col1 ) (
    partition p1 values less than (11),
    partition p2 values less than (15)
);
```

**Comment by tiancaiamao** [comment]
No matter it's a primary key, or unique key, or multiple unique key, it can be

```
type UniqueKey map[string]struct{}
```

Then the logic is very simple:

```
// check returns true if union u1 >= u2.
func checkOne(u1, u2 UniqueKey) {
     for _, v := range u2 {
         if _, ok := u1[v]; !ok {
            return false
         }
     }
     return true
}

func check(constrains []UniqueKey, partition UniqueKey) {
     for _, constrain := range constrains {
         if !checkOne(constrain, partition) {
            return false
         }
     }
     return true
}
```

File: ddl/partition.go [hunk 7]
```diff
@@ -280,3 +281,96 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+		}
+		if mysql.HasPriKeyFlag(col.Flag) {
+			priKeys[col.Name.L] = struct{}{}
+		}
+	}
+	return uniKeys, multipleKeys, priKeys
+}
+
+// checkConstraintIncludePartKey checks that the partitioning key is included in the constraint.
+func checkConstraintIncludePartKey(partkeys []string, constraints map[string]struct{}, multipleColumn bool) bool {
```
Comment (tiancaiamao) [inline ddl/partition.go]: > checkConstraintIncludePartKey checks that the partitioning key is included in the constraint.

I don't think we need the `multipleColumn bool` parameter:

```
func checkConstraintIncludePartKey(partkeys []string, constraints map[string]struct{}) {
    for _, pk := range partKeys {
        if _, ok := constraints[pk]; !ok {
            return false
        }
    }
    return true
}
```

File: ddl/partition.go [hunk 8]
```diff
@@ -305,72 +305,62 @@ func checkRangePartitioningKeysConstraints(ctx sessionctx.Context, expr ast.Expr
...
 	}
-	return uniKeys, multipleKeys, priKeys
+	return constraints
 }
 
 // checkConstraintIncludePartKey checks that the partitioning key is included in the constraint.
-func checkConstraintIncludePartKey(partkeys []string, constraints map[string]struct{}, multipleColumn bool) bool {
-	count := 0
+func checkConstraintIncludePartKey(partkeys []string, constraints map[string]struct{}) bool {
 	if len(constraints) > 0 {
```
Comment (tiancaiamao) [inline ddl/partition.go]: Remove this if:

If len(constraints) == 0, function `checkConstraintIncludePartKey` should return false.

File: ddl/partition.go [hunk 9]
```diff
@@ -305,72 +305,62 @@ func checkRangePartitioningKeysConstraints(ctx sessionctx.Context, expr ast.Expr
...
 	}
 	return nil
 }
 
-// saveConstraintsColumnNames save the column names in the table constraint in map.
-func saveConstraintsColumnNames(tblInfo *model.TableInfo, constraints []*ast.Constraint) (map[string]struct{}, map[string]struct{}, map[string]struct{}) {
-	multipleKeys := make(map[string]struct{})
-	for _, v := range constraints {
+// saveConstraintsColumnNames Save the column names in table constraints to []map[string]struct{}.
+func saveConstraintsColumnNames(tblInfo *model.TableInfo, cons []*ast.Constraint) []map[string]struct{} {
```
Comment (tiancaiamao) [inline ddl/partition.go]: Could this function handle : 
```
unique key (col1, col2)
unique key (col1, col3)
unique key col4
primary key col5
```

and return 
```
[]map[string]struct{} {
    map[string]struct{} {  "col1" : struct{}{}, "col2" : struct{}{}   },
    map[string]struct{} {  "col1" : struct{}{}, "col3" : struct{}{}   },
    map[string]struct{} {  "col4" : struct{}{}  },
    map[string]struct{} {  "col5" : struct{}{}  },
}
```

File: ddl/partition.go [hunk 10]
```diff
@@ -305,72 +305,62 @@ func checkRangePartitioningKeysConstraints(ctx sessionctx.Context, expr ast.Expr
...
-		if checkConstraintIncludePartKey(partkeys, priKeys, false) || checkConstraintIncludePartKey(partkeys, multipleKeys, false) {
-			return ErrUniqueKeyNeedAllFieldsInPf.GenByArgs("PRIMARY KEY")
-		}
-	default:
-		// Other expression, with multiple column names.
-		if checkConstraintIncludePartKey(partkeys, priKeys, true) || checkConstraintIncludePartKey(partkeys, multipleKeys, true) {
-			return ErrUniqueKeyNeedAllFieldsInPf.GenByArgs("PRIMARY KEY")
-		}
-
+	// Checks that the partitioning key is included in the constraint.
```
Comment (tiancaiamao) [inline ddl/partition.go]: That's enough for the rest of the function:
```
if !checkConstraintIncludePartKey(partKeys, consColNames) {
    return ErrUniqueKeyNeedAllFieldsInPf
}
return nil
```
We don't care for the error message, as long as the error code is right.

File: ddl/partition.go [hunk 11]
```diff
@@ -305,72 +305,62 @@ func checkRangePartitioningKeysConstraints(ctx sessionctx.Context, expr ast.Expr
...
+	if !checkConstraintIncludePartKey(partkeys, consColNames[0]) || !checkConstraintIncludePartKey(partkeys, consColNames[1]) {
+		return ErrUniqueKeyNeedAllFieldsInPf.GenByArgs("PRIMARY KEY")
 	}
 
-	// Every unique key on the table must use every column in the table's partitioning expression
-	// see https://dev.mysql.com/doc/refman/5.7/en/partitioning-limitations-partitioning-keys-unique-keys.html.
-	if len(uniKeys) > 0 {
-		_, ok := uniKeys[partkeys[0]]
-		if len(uniKeys) > 1 || len(partkeys) > 1 || !ok {
+	if len(consColNames[2]) > 0 {
```
Comment (tiancaiamao) [inline ddl/partition.go]: Remove this

File: ddl/partition.go [hunk 12]
```diff
@@ -280,3 +281,88 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+		return nil
+	}
+	_, ok := consColNames[2][partkeys[0]]
+	if len(consColNames[2]) > 1 || len(partkeys) > 1 || !ok {
+		return ErrUniqueKeyNeedAllFieldsInPf
+	}
+	return nil
+}
+
+// saveConstraintsColumnNames Save the column names in table constraints to []map[string]struct{}.
```
Comment (shenli) [inline ddl/partition.go]: s/Save/saves/

File: ddl/partition.go [hunk 13]
```diff
@@ -280,3 +281,88 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cols := expression.ExtractColumns(e)
+	for _, col := range cols {
+		partkeys = append(partkeys, col.ColName.L)
+	}
+
+	// Checks that the partitioning key is included in the constraint.
+	// Here consColNames[0] saves Multiple keys, consColNames[1] saves Primary keys.
```
Comment (shenli) [inline ddl/partition.go]: saves?

File: ddl/partition.go [hunk 14]
```diff
@@ -280,3 +281,88 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+	if !checkConstraintIncludePartKey(partkeys, consColNames[0]) || !checkConstraintIncludePartKey(partkeys, consColNames[1]) {
+		return ErrUniqueKeyNeedAllFieldsInPf
+	}
+
+	// Every unique key on the table must use every column in the table's partitioning expression.
+	//  Here consColNames[2] saves Unique key.
```
Comment (shenli) [inline ddl/partition.go]: Remove the extra whitespace before `Here`.

File: ddl/partition.go [hunk 15]
```diff
@@ -280,3 +281,88 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+	// see https://dev.mysql.com/doc/refman/5.7/en/partitioning-limitations-partitioning-keys-unique-keys.html.
+	if len(consColNames[2]) == 0 {
+		return nil
+	}
+	_, ok := consColNames[2][partkeys[0]]
+	if len(consColNames[2]) > 1 || len(partkeys) > 1 || !ok {
```
Comment (shenli) [inline ddl/partition.go]: Why not put `ok` at the first place?

File: ddl/partition.go [hunk 16]
```diff
@@ -280,3 +281,88 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
 	}
 	return partitionIDs
 }
+
+// checkRangePartitioningKeysConstraints checks that the range partitioning key is included in the table constraint.
+func checkRangePartitioningKeysConstraints(ctx sessionctx.Context, expr ast.ExprNode, tblInfo *model.TableInfo, constraints []*ast.Constraint) error {
```
Comment (shenli) [inline ddl/partition.go]: How could you tell it is a range partitioned table?

**Comment by ciscoxll** [comment]
@shenli @tiancaiamao @winkyao PTAL.

File: ddl/partition.go [hunk 17]
```diff
@@ -280,3 +281,88 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
 	}
 	return partitionIDs
 }
+
+// checkRangePartitioningKeysConstraints checks that the range partitioning key is included in the table constraint.
+func checkRangePartitioningKeysConstraints(ctx sessionctx.Context, s *ast.CreateTableStmt, tblInfo *model.TableInfo, constraints []*ast.Constraint) error {
+	// Returns directly if there is no constraint in the partition table.
+	if s.Partition.Tp != model.PartitionTypeRange || len(constraints) == 0 {
```
Comment (shenli) [inline ddl/partition.go]: Please move the "s.Partition.Tp != model.PartitionTypeRange" out of the function. It should be checked before entering the function.

File: ddl/partition.go [hunk 18]
```diff
@@ -280,3 +281,88 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+	}
+	_, ok := consColNames[2][partkeys[0]]
+	if !ok || len(consColNames[2]) > 1 || len(partkeys) > 1 {
+		return ErrUniqueKeyNeedAllFieldsInPf
+	}
+	return nil
+}
+
+// saveConstraintsColumnNames saves the column names in table constraints to []map[string]struct{}.
+func saveConstraintsColumnNames(tblInfo *model.TableInfo, cons []*ast.Constraint) []map[string]struct{} {
```
Comment (shenli) [inline ddl/partition.go]: s/saveConstraintsColumnNames/extractConstraintsColumnNames/

File: ddl/partition.go [hunk 19]
```diff
@@ -280,3 +281,88 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cols := expression.ExtractColumns(e)
+	for _, col := range cols {
+		partkeys = append(partkeys, col.ColName.L)
+	}
+
+	// Checks that the partitioning key is included in the constraint.
+	// Here consColNames[0] save Multiple keys, consColNames[1] save Primary keys.
```
Comment (shenli) [inline ddl/partition.go]: s/save/contains/

File: ddl/partition.go [hunk 20]
```diff
@@ -280,3 +281,88 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+	if !checkConstraintIncludePartKey(partkeys, consColNames[0]) || !checkConstraintIncludePartKey(partkeys, consColNames[1]) {
+		return ErrUniqueKeyNeedAllFieldsInPf
+	}
+
+	// Every unique key on the table must use every column in the table's partitioning expression.
+	// Here consColNames[2] saves Unique key.
```
Comment (shenli) [inline ddl/partition.go]: s/saves/contains/

File: ddl/partition.go (hunk 18 above)
Comment (shenli) [reply ddl/partition.go]: It is better to return three `map[string]struct{}`.

File: ddl/partition.go [hunk 21]
```diff
@@ -280,3 +281,88 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+	}
+	return nil
+}
+
+// saveConstraintsColumnNames saves the column names in table constraints to []map[string]struct{}.
+func saveConstraintsColumnNames(tblInfo *model.TableInfo, cons []*ast.Constraint) []map[string]struct{} {
+	multipleKeys := make(map[string]struct{})
+	for _, v := range cons {
+		if v.Tp == ast.ConstraintUniq {
+			for _, key := range v.Keys {
```
Comment (shenli) [inline ddl/partition.go]: Is primary-key also in the v.Keys?

**Comment by ciscoxll** [comment]
@tiancaiamao @winkyao @shenli PTAL

File: ddl/db_test.go [hunk 22]
```diff
@@ -595,7 +595,7 @@ func (s *testDBSuite) TestAddMultiColumnsIndex(c *C) {
 func (s *testDBSuite) TestAddIndex(c *C) {
 	s.testAddIndex(c, false, "create table test_add_index (c1 bigint, c2 bigint, c3 bigint, primary key(c1))")
 	s.testAddIndex(c, true, `create table test_add_index (c1 bigint, c2 bigint, c3 bigint, primary key(c1))
-			      partition by range (c2) (
+			      partition by range (c1) (
```
Comment (shenli) [inline ddl/db_test.go]: Why change this?

Comment (ciscoxll) [reply ddl/db_test.go]: @shenli This SQL is wrong, It was checked out by this PR.

File: ddl/partition.go [hunk 23]
```diff
@@ -280,3 +283,87 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+	return nil
+}
+
+// extractConstraintsColumnNames extract the column names in table constraints to []map[string]struct{}.
+func extractConstraintsColumnNames(tblInfo *model.TableInfo, cons []*ast.Constraint) []map[string]struct{} {
+	var constraints []map[string]struct{}
+	for _, v := range cons {
+		if v.Tp == ast.ConstraintUniq {
+			multipleKeys := make(map[string]struct{})
+			for i, key := range v.Keys {
```
Comment (tiancaiamao) [inline ddl/partition.go]: How about
```
tmp := make(map[string]struct{})
for _, key := range v.Keys {
    tmp[key.Column.Name.L] = struct{}{}
}
constraints = append(constraints, multipleKeys)
```
No matter it's a single unique key or multiple unique keys, it's collected here.

Then remove line 343~349

**Comment by ciscoxll** [comment]
@shenli @tiancaiamao @winkyao PTAL.

File: ddl/partition.go [hunk 24]
```diff
@@ -280,3 +281,79 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+
+// extractConstraintsColumnNames extract the column names in table constraints to []map[string]struct{}.
+func extractConstraintsColumnNames(cons []*ast.Constraint) []map[string]struct{} {
+	var constraints []map[string]struct{}
+	for _, v := range cons {
+		if v.Tp == ast.ConstraintUniq {
+			uniKeys := make(map[string]struct{})
+			for i, key := range v.Keys {
+				uniKeys[key.Column.Name.L] = struct{}{}
+				if i == len(v.Keys) -1 {
```
Comment (tiancaiamao) [inline ddl/partition.go]: This can be more out of this for loop

File: ddl/partition.go [hunk 25]
```diff
@@ -280,3 +281,79 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+						constraints = append(constraints, uniKeys)
+					}
+				}
+			}
+		}
+		if v.Tp == ast.ConstraintPrimaryKey {
+			priKeys := make(map[string]struct{})
+			for j, key := range v.Keys {
+				priKeys[key.Column.Name.L] = struct{}{}
+				if j == len(v.Keys) -1 {
```
Comment (tiancaiamao) [inline ddl/partition.go]: ditto

**Comment by tiancaiamao** [comment]
LGTM

**Comment by ciscoxll** [comment]
@shenli @zimulala  @winkyao PTAL.

**Comment by ciscoxll** [comment]
@winkyao PTAL.

File: ddl/partition.go [hunk 26]
```diff
@@ -280,3 +281,76 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+		}
+	}
+	return nil
+}
+
+// extractConstraintsColumnNames extract the column names in table constraints to []map[string]struct{}.
+func extractConstraintsColumnNames(cons []*ast.Constraint) []map[string]struct{} {
+	var constraints []map[string]struct{}
+	for _, v := range cons {
+		if v.Tp == ast.ConstraintUniq {
```
[the remark this answers is not in the document]
Comment (ciscoxll) [reply ddl/partition.go]: @winkyao This change will result in an extract every primary key error.

[the remark this answers is not in the document]
Comment (ciscoxll) [reply ddl/partition.go]: @winkyao Done.

**Pull request merged**